File: src/main/kotlin/io/titandata/titan/providers/kubernetes/Checkout.kt
/*
 * Copyright (c) 2019 by Delphix. All rights reserved.
 */

package io.titandata.titan.providers.kubernetes

import io.titandata.client.apis.CommitsApi
import io.titandata.client.apis.RepositoriesApi
import io.titandata.client.apis.VolumesApi
import io.titandata.titan.clients.Kubernetes

class Checkout(
    private val kubernetes: Kubernetes = Kubernetes(),
    private val commitsApi: CommitsApi = CommitsApi(),
    private val repositoriesApi: RepositoriesApi = RepositoriesApi(),
    private val volumesApi: VolumesApi = VolumesApi()
) {
    fun checkout(repoName: String, guid: String?, tags: List<String>) {
        val sourceCommit = if (guid == null) {
            if (tags.isNotEmpty()) {
                val commits = commitsApi.listCommits(repoName, tags)
                if (commits.size == 0) {
                    throw IllegalStateException("no matching commits found")
                }
                commits.first().id
            } else {
                val status = repositoriesApi.getRepositoryStatus(repoName)
                if (status.sourceCommit == null) {
                    throw IllegalStateException("no commits present, run 'titan commit' first")
                }
                status.sourceCommit!!
            }
        } else {
            if (tags.isNotEmpty()) {
                throw IllegalArgumentException("tags and commit cannot both be specified")
            }
            guid
        }

        val repo = repositoriesApi.getRepository(repoName)

        val status = commitsApi.getCommitStatus(repoName, sourceCommit)
        if (!status.ready) {
            println("Waiting for commit to be ready")
            while (true) {
                val commitStatus = commitsApi.getCommitStatus(repoName, sourceCommit)
                if (commitStatus.ready) {
                    break
                }
                Thread.sleep(1000L)
            }
        }

        println("Checkout $sourceCommit")
        commitsApi.checkoutCommit(repoName, sourceCommit)

        println("Stopping port forwarding")
        kubernetes.stopPortFowarding(repoName)

        println("Updating deployment")
        kubernetes.updateStatefulSetVolumes(repoName, volumesApi.listVolumes(repoName).toList())

        println("Waiting for deployment to be ready")
        kubernetes.waitForStatefulSet(repoName)

        println("Starting port forwarding")
        kubernetes.startPortForwarding(repoName)
    }
}

File: internal/app/providers/kubernetes/Checkout.go
package kubernetes

import (
	"fmt"
	"github.com/antihax/optional"
	titanclient "github.com/titan-data/titan-client-go"
	"os"
	"strconv"
	"time"
)

func Checkout(repoName string, guid string, tags []string, port int)  {
	cfg.BasePath = "http://localhost:" + strconv.Itoa(port)

	var sourceCommit string
	if guid == "" {
		if len(tags) > 0 {
			o := optional.NewInterface(tags)
			l := titanclient.ListCommitsOpts{Tag: o}
			commits, _, _ := commitsApi.ListCommits(ctx, repoName, &l)
			if len(commits) == 0 {
				fmt.Println("no matching commits found")
				os.Exit(1)
			}
			sourceCommit = commits[0].Id
		} else {
			status, _, _ := repositoriesApi.GetRepositoryStatus(ctx, repoName)
			if status.SourceCommit == "" {
				fmt.Println("no commits present, run 'titan commit' first")
				os.Exit(1)
			}
			sourceCommit = status.SourceCommit
		}

	} else {
		if len(tags) > 0 {
			fmt.Println("tags and commit cannot both be specified")
			os.Exit(1)
		}
		sourceCommit = guid
	}

	status, _, _ := commitsApi.GetCommitStatus(ctx, repoName, sourceCommit)

	if !status.Ready {
		fmt.Println("Waiting for commit to be ready")
		c := true
		for c {
			commitStatus, _, _ := commitsApi.GetCommitStatus(ctx, repoName, sourceCommit)
			if commitStatus.Ready {
				c = false
			}
			time.Sleep(1000)
		}
	}

	fmt.Println("Checkout " + sourceCommit)
	commitsApi.CheckoutCommit(ctx, repoName, sourceCommit)

	fmt.Println("Stopping port forwarding")
	k8s.StopPortForwarding(repoName)

	fmt.Println("Updating deployment")
	vols, _, _ := volumesApi.ListVolumes(ctx, repoName)
	k8s.UpdateStatefulSetVolumes(repoName, vols)

	fmt.Println("Waiting for deployment to be ready")
	k8s.WaitForStatefulSet(repoName)

	fmt.Println("Starting port forwarding")
	k8s.StartPortForwarding(repoName)
}